### ISSUE TITLE
Integration with Piranha CMS

### ISSUE THREAD

**Issue by tidyui** [simplcommerce/SimplCommerce#426]
Hi there @thiennn! I'm the lead developer for @PiranhaCMS, an open source CMS for .NET that's been around since 2011. In december last year I released the completely rewritten version for .NET core (`4.0`) and I'm currently working on version `4.3`.

I would be really interested in providing a good integration with you E-commerce solution as this would be a perfect fit for both sides :) I haven't had time yet to look at the architecture of SimplCommerce, but if you think it sounds interesting I'd be more than happy to have a discussion regarding how we could integrate the systems in the best possible way!

Regards!

**Comment by psijkof** [comment]
Yes, we had the same thoughts, and this is great that @tidyui is proposing this!

**Comment by tidyui** [comment]
Yes, I'm waiting for @thiennn to respond before taking any action though as it's probably better to discuss briefly how to do it in the best way before getting started.

**Comment by thiennn** [comment]
Sorry I haven't had time to look at PiranhaCMS and have no idea how it should be integrated.

**Comment by tidyui** [comment]
I understand that, I have only had the time to glance at the code for SimpleCommerce, but for starters.

## Permalinks

By looking at your demo site I can see that you intercept requests for (as an example):

* `/category-slug`
* `/product-slug`
* `/cart`
* `/login`
* `/register`

and probably dozens of other pages I'm not including here. The main pages for browsing the shop itself is obviously `/category-slug` & `/product-slug`.

A key factor to make a nice integration with any CMS would probably be to be able to configure a root **prefix** for the routing of SimplCommerce, for example positioning all routes below `/products/` or `/shop/`. This might be possible already, but if this exists then it would be easy to put the rest of the website content on other routes.

Also, is the web application part of **SimplCommerce** placed in an `Area`, or do we run the risk of having name collisions on **shared views** when running it in the same application as the CMS?

Just some starting questions that came to mind :)

**Comment by tidyui** [comment]
And to clarify how Piranha works. The entire routing is basically just a couple of `middleware` components that checks the incoming requests, and if they exist in the CMS it just **rewrites** the request to the appropriate `route`. After that it's just plain MVC

**Comment by praveensewak** [comment]
Hello gents! Any movement with this? I come from NopCommerce background, where I have developed a custom CMS (private unfortunately) on top of it. I have another big project coming up, and been looking around and this 'marriage' would be awesome solution for it. Happy to contribute once the ball gets rolling.

**Comment by tidyui** [comment]
I haven’t received any feedback so I guess I’ll just have to try it out. Problem is I’m knee deep in finishing up the 4.4 release, but after that I could absolutely put some effort into this!

**Comment by thiennn** [comment]
In SimplCommere, routing can be categorized into 2 groups:
 - Static routes: archived by adding Route attribute on Actions or Controllers of fall to default route
 - Dynamic routes: such as categories, products,..... archived by https://github.com/simplcommerce/SimplCommerce/blob/master/src/Modules/SimplCommerce.Module.Core/Extensions/UrlSlugRoute.cs

The dynamic route is executed first, if there are no item matched then it will continue to go to static routes, if static routes not matched then fallback to default.

**Comment by tidyui** [comment]
Thank. Like I said I'd be more than happy to try it out. However I'd like to know some more from @praveensewak & @psijkof how an ideal integration would look like. Is it sufficient to have Piranha & SimplCommerce running in the same application, with SimplCommerce on a certain route prefix, and have two different admin interfaces? 

Or are you hoping to achieve a tighter integration, like presenting products from SimplCommerce in a Piranha page (or the other way around, loading up CMS-data from Piranha into the SimplCommerce pages).

Regards

**Comment by bendandoSolo** [comment]
Just started using Piranha and really like it, would be great to have an integrated Ecommerce platform, please let me know if this progresses

**Comment by hishamco** [comment]
@tidyui @thiennn the integration will be cool, but:

1- How the integration will be
2- What could be integrated

but I think Piranha CMS may use SimplCommerce as e-Commerce module

**Comment by dinukasal** [comment]
@praveensewak @tidyui I am also going to integrate nopcommerce and piranha. Did any update on this discussion happen?

**Comment by tnlthanzeel** [comment]
@tidyui , hi. we are about to use piranha cms in our company. is it possible to add something like shopify?
